### PR TITLE
Add report/unreport to post & comment, discussion needed

### PR DESCRIPTION
**Pull request by Henry-Graves**
# Describe your changes
Made xplat functions addReport, removeReport, reportedBy (to check if given user has reported the content).
Reports are stored in a new Firebase collection. 
I didn't make a new type for reports but can change that if it's a big deal.

@JakeSteinebronn @tylerhm @abrugal  For auto-hiding content after 3 reports I'm gonna add a boolean "hiddenByReports". Do want to filter out that content when fetching or client-side? I think it should be a front-end check.

# Issue ticket number and link
Finishes some sub-tasks for #105

Next up: auto-hide content, employee actions.

### CODE REVIEW

**Comment by JakeSteinebronn** [comment]
for auto-hiding, we can have the report transaction also increment a counter on every post like `numReports` (and when a report is resolved, it decrements this count)

Then, front-end can just (JUST) check if that number is gequal to some const, like 3 or 5 or whatever.

What do you think

**Comment by Henry-Graves** [comment]
That would work, the report list is currently a field on every post & comment so front-end can check the size of the list.
It saves us from storing an extra bool on every post/comment, but is probably a bit more annoying to use. I think I'd rather keep an updated bool

**Comment by JakeSteinebronn** [comment]
Well you wouldn't ever need the bool, you could have your `checkShouldBeHidden` just do the count / length check. I also think we don't need to maintain the actual array of the reports, just the reports themselves

**Review by JakeSteinebronn — CHANGES_REQUESTED** [review]
I think this can be considerably cleaner if we move it into user. something like

`User::reportContent(content: Post | Comment | User) {}`
Since really, all you need is the docref of the content, and they all extend lazy object, and you can just add the `reportedBy[]` list to these 3 types and it'll all work out.

**Comment by Henry-Graves** [comment]
I moved the redundant functions to User, but didn't opt to support reporting a User itself, it'll keep reporting simpler for employees because they won't see a reported user then have to dig through all their content to find why they're reported.

added checkShouldBeHidden() to Post and Comment

also simplified the code for the functions by checking if they have data & getting data at the top of the functions

**Comment by JakeSteinebronn** [comment]
We can get to this later, but user reporting should be implemented if their avatar, displayname, or bio is inappropriate

**Comment by Henry-Graves** [comment]
oh true I'll add that next then

File: types/user.ts [hunk 1]
```diff
@@ -85,6 +89,70 @@ export class User extends LazyObject {
...
+    // update server side
+    const q = await getDocs(
+      query(
+      collection(db, 'reports'),
+      where('reporter', '==', this.docRef!),
+      where('content', '==', content.docRef!)
+      )
+    );
+    const reportDocRef = q.docs[0].ref;
+    return runTransaction(db, async (transaction: Transaction) => {
```
Comment (JakeSteinebronn) [inline types/user.ts:142]: Don't we also have to update `content`'s `reports` array to remove this user docref?

File: types/user.ts [hunk 2]
```diff
@@ -85,6 +89,70 @@ export class User extends LazyObject {
...
+    // if already reported, return
+    if (await this.alreadyReported(content)) return;
+
+    // update client side
+    content.reports?.push(this);
+
+    // update server side
+    const newReportDocRef = doc(collection(db, 'reports'));
+
+    return runTransaction(db, async (transaction: Transaction) => {
```
Comment (JakeSteinebronn) [inline types/user.ts:108]: Don't we have to add to `content`'s `reports` array ?

**Review by JakeSteinebronn — CHANGES_REQUESTED** [review]

**Review by JakeSteinebronn — APPROVED** [review]

**Comment by tylerhm** [comment]
> Do want to filter out that content when fetching or client-side?

Front-end check sounds good to me :)

**Pull request merged**